File: ruby/lib/models/teller.rb
class Teller

  def initialize(catalog)
    @catalog = catalog
    @offers = {}
  end

  def add_special_offer(offer_type, product, argument)
    @offers[product] = Offer.new(offer_type, product, argument)
  end

  def checks_out_articles_from(the_cart)
    receipt = Receipt.new
    product_quantities = the_cart.items
    for pq in product_quantities do
      p = pq.product
      quantity = pq.quantity
      unit_price = @catalog.unit_price(p)
      price = quantity * unit_price
      receipt.add_product(p, quantity, unit_price, price)
    end
    the_cart.handle_offers(receipt, @offers, @catalog)

    receipt
  end

end

File: elixir/lib/supermarket/model/teller.ex
defmodule Supermarket.Model.Teller do
  alias Supermarket.Model.Offer
  alias Supermarket.Model.Receipt
  alias Supermarket.Model.ShoppingCart
  alias Supermarket.Model.SupermarketCatalog

  defstruct [:catalog, :offers]

  def new(catalog) do
    %__MODULE__{catalog: catalog, offers: %{}}
  end

  def add_special_offer(teller, offer_type, product, argument) do
    Map.update!(teller, :offers, &Map.put(&1, product, Offer.new(offer_type, product, argument)))
  end

  def checks_out_articles_from(teller, the_cart) do
    receipt = Receipt.new()
    product_quantities = the_cart.items

    receipt =
      product_quantities
      |> Enum.reverse()
      |> Enum.reduce(receipt, fn pq, receipt ->
        p = pq.product
        quantity = pq.quantity
        unit_price = SupermarketCatalog.get_unit_price(teller.catalog, p)
        price = quantity * unit_price
        Receipt.add_product(receipt, p, quantity, unit_price, price)
      end)

    ShoppingCart.handle_offers(the_cart, receipt, teller.offers, teller.catalog)
  end
end
